Document process for registering a plugin as a Packer Integration (#12702)
hashicorp/packer: website/content/docs/plugins/creation/index.mdx

   releaser is working. The tag must be a valid
   [Semantic Version](https://semver.org/) preceded with a `v`. Once the tag is pushed, the github actions you just configured will automatically build release binaries that Packer can download using `packer init`. For more details on how
   to install a plugin using `packer init`, see the
   [init docs](https://developer.hashicorp.com/packer/docs/commands/init).

## Registering Plugins

> ~> Note: Registering a plugin as an integration requires the use of [Packer's Incofiguration configuration](https://github.com/hashicorp/packer-plugin-scaffolding/tree/main/docs).

Reviewer: Not sure what the exact intent was here, but I believe we only require the final hierarchy to match what's in `.web-docs`, how a developer structures their code doesn't matter, so long as we can get this from the git repo.

```suggestion
~> Note: Registering a plugin as an integration requires the documentation to match the [Scaffolding example layout](https://github.com/hashicorp/packer-plugin-scaffolding/tree/main/.web-docs).
```

The link will be dead until we merge https://github.com/hashicorp/packer-plugin-scaffolding/pull/108 too

Reviewer: Also, since we state below that the directory structure must respect a convention (where do we document it?), we could also do away with this note, and add the link to the scaffolding's `.web-docs` in here

> ~> Note: Registering a plugin as an integration requires the documentation to match the [Scaffolding example layout](https://github.com/hashicorp/packer-plugin-scaffolding/tree/main/.web-docs).
> 

Author: I think we don't need this note given the recent suggestion, which is what I think you were alluding to @lbajolet-hashicorp 
```suggestion

```

Reviewer: Spot on!

To help with the discovery of Packer plugins, plugins maintainers can choose to register their plugin as a [Packer Integration](https://developer.hashicorp.com/packer/integrations).

> The registration process requires [metadata configuration](https://github.com/hashicorp/integration-template#metadata-configuration) be added to your plugin repository for configuring the Packer integration pipeline and 

Reviewer 2: ```suggestion
The registration process requires [metadata configuration](https://github.com/hashicorp/integration-template#metadata-configuration) to be added to your plugin repository for configuring the Packer integration pipeline and 
```

> The registration process requires [metadata configuration](https://github.com/hashicorp/integration-template#metadata-configuration) be added to your plugin repository for configuring the Packer integration pipeline and 

Reviewer: ```suggestion
The registration process requires [metadata configuration](https://github.com/hashicorp/integration-template#metadata-configuration) to be added to your plugin repository for configuring the Packer integration pipeline and 
```

> a specific directory structure for plugin documentation to be render on the [Packer Integrations](https://developer.hashicorp.com/packer/integrations) portal.

Reviewer: ```suggestion
a specific directory structure for plugin documentation to be rendered on the [Packer Integrations](https://developer.hashicorp.com/packer/integrations) portal.
```

> The registration process requires [metadata configuration](https://github.com/hashicorp/integration-template#metadata-configuration) to be added to your plugin repository for configuring the Packer integration pipeline and 
> a specific directory structure for plugin documentation to be rendered on the [Packer Integrations](https://developer.hashicorp.com/packer/integrations) portal.

Author: ```suggestion
The registration process requires [metadata configuration](https://github.com/hashicorp/integration-template#metadata-configuration) to be added to your plugin repository for configuring the Packer integration pipeline and a `web-docs` directory structure for plugin documentation to be rendered on the [Packer Integrations](https://developer.hashicorp.com/packer/integrations) portal.
```

Reviewer: Updating your suggestion to point to the `.web-docs` example from scaffolding

You can execute the following steps to register your plugin as an integration:

> 1. Update your plugin documentation structure according to the template defined with the [Packer Plugin Scaffolding template].

Reviewer 2: ```suggestion
1. Update your plugin documentation structure according to the template defined with the [Packer Plugin Scaffolding template](https://github.com/hashicorp/packer-plugin-scaffolding/tree/main/docs).
```

Author: @lbajolet-hashicorp I updated the suggestion. Let me know what you think?

Reviewer: I'm not too sure about the `Plugins using `packer-sdc` for generating documentation partials should look to create a new directory for serving the fully rendered documentation.`
`packer-sdc` is built-in and required to generate the hcl2spec Go files for the structures exchanged over-the-wire, so I think it's reasonable to assume devs are using it.
I understand they may not rely on it for documentation purposes, but that's minor I'd think.

Author: Okay I'll remove that wording

> New plugins generated from this template may have the necessary structure in place. If so you can jump to step 3.

Reviewer: ```suggestion
New plugins generated from this template have the necessary structure in place. If so you can jump to step 3.
```

1. For the integrations library, only one top-level README per integration is supported. Any top-level index.mdx files that exist

> within a plugins existing documentation will need to migrate to a top-level README. 

Reviewer: ```suggestion
within a plugin's existing documentation will need to migrate to a top-level README. 
```

1. Update your top-level integration README to include a description, plugin installation steps, available components section, and, any, additional sections
needed to inform users on how to work with your integration. Refer to [Packer scaffolding plugin](https://github.com/hashicorp/packer-plugin-scaffolding/blob/main/docs/README.md) for an example. 
1. Update the top-level README for each of the components within your integration to follow the structure defined in the scaffolding template.
1. Add the integration configuration file [metadata.hcl](https://github.com/hashicorp/packer-plugin-scaffolding/blob/main/.web-docs/metadata.hcl) to the plugins `.web-docs` directory.

> 1. Open a request for integration issue with the Packer team - [Open Request](https://github.com/hashicorp/packer/issues/new/choose).

Author: Yes - it is part of this PR.

> 1. Open a request for integration issue with the Packer team - [Open Request](https://github.com/hashicorp/packer/issues/new/choose).

Reviewer: We can link directly to the appropriate template for this I'd think

```suggestion
1. Open a request for integration issue with the Packer team - [Open Request](https://github.com/hashicorp/packer/issues/new?labels=new-plugin-contribution&template=plugin_integration.md).
```

> Please provide all the requested information to help expedite the integration request. 

Reviewer 2: I remember a discussion about not using "please" on docs and error messages.
```suggestion
Provide all the requested information to help expedite the integration request. 
```

#### [Example] Add integration files to existing plugin repository

```shell
## Update Plugin repository with integration config, workflows, and scripts
cd packer-plugin-name
mkdir -p .web-docs/scripts

# Download packer-plugin-scaffolding repo copy files

> curl -L -O https://github.com/hashicorp/packer-plugin-scaffolding/archive/refs/heads/main.zip

Reviewer: I know how ubiquitous `curl` is, but `wget` is even more, and doesn't need flags to do just what we need here :)

```suggestion
wget https://github.com/hashicorp/packer-plugin-scaffolding/archive/refs/heads/main.zip
```

cp packer-plugin-scaffolding-main/.web-docs/metadata.hcl .web-docs/
cp -r packer-plugin-scaffolding-main/.web-docs/scripts/ .web-docs/scripts/
cp .github/workflows/notify-integration-release-via-* .github/workflows/

# Remove downloaded scaffolding project
rm main.zip
rm -rf packer-plugin-scaffolding-main

# Add the following commands to your plugin GNUmakefile
generate: install-packer-sdc
    @go generate ./...
    @rm -rf .docs
    @packer-sdc renderdocs -src docs -partials docs-partials/ -dst .docs/
    @./.web-docs/scripts/compile-to-webdocs.sh "." ".docs" ".web-docs" "<orgname>"
    @rm -r ".docs"
```

By opening an integration request, you are asking a member of the to Packer team to review your plugin integration configuration, plugin documentation, 
and, finally, to open an internal pull-request to finalize the integration setup. 

Plugin integrations will be listed as a [Packer Integration](https://developer.hashicorp.com/packer/integrations), with details on how to install and use your the plugin. 

Plugin integrations, once deployed, can be updated manually, or automatically upon a new release, by the plugin authors. Changes to the defined documentation structure 

> or parent repository should be communicated to the Packer team to ensure a working integration pipeline. 

Reviewer 2: how should the communication be? through which channel?

Author: 🤷‍♂️ - good question. I'll suggest opening an issue or sending an email to the packer integration distro.

Reviewer 2: Maybe add the information here as well

## Registering Plugins

~> Note: Registering a plugin as an integration requires the documentation to match the [Scaffolding example layout](https://github.com/hashicorp/packer-plugin-scaffolding/tree/main/.web-docs).

To help with the discovery of Packer plugins, plugins maintainers can choose to register their plugin as a [Packer Integration](https://developer.hashicorp.com/packer/integrations).

The registration process requires [metadata configuration](https://github.com/hashicorp/integration-template#metadata-configuration) to be added to your plugin repository for configuring the Packer integration pipeline and 
a specific directory structure for plugin documentation to be rendered on the [Packer Integrations](https://developer.hashicorp.com/packer/integrations) portal.

You can execute the following steps to register your plugin as an integration:

1. Update your plugin documentation structure according to the template defined with the [Packer Plugin Scaffolding template].

> New plugins generated from this template have the necessary structure in place. If so you can jump to step 3.

Author: ```suggestion
New plugins generated from the scaffolding template have the necessary structure in place. If so you can jump to step 3.
```

   to install a plugin using `packer init`, see the
   [init docs](/packer/docs/commands/init).

## Plugin Integration Registration

~> Note: Registering a remote as an integration requires the use of [Packer's plugin docs configuration](https://github.com/hashicorp/packer-plugin-scaffolding/tree/main/docs).

> `packer init` allows users to require and install remote Packer plugins published to GitHub, those not bundled with Packer core, automatically.

Reviewer: Since we won't have bundled plugins soon, we can remove this part :)

To help with the discovery of remote Packer plugins, plugins maintainers can choose to register their plugin as a [Packer Integration](/packer/integrations).

> The registration process requires changes to your plugin repository for configuring the Packer integration pipeline and plugin documentation for rendering 

Reviewer: Should we say that the registration requires changes here? This is the case for plugin authors that adopted the scaffolding before this change, but once we've updated the structure, this won't apply, as the new structure will be in place.

on the [Packer Integrations](/packer/integrations) portal.

> Once in place an integration request can be made on Packer GitHub repository. By opening an integration request you are asking a member of the to Packer

Reviewer: ```suggestion
Once in place, an integration request can be made on Packer GitHub repository. By opening an integration request, you are asking a member of the to Packer
```

team to review your plugin integration configuration, plugin documentation, and, finally, to open an internal pull-request to finalize the integration setup. 

Once completed your the plugin integration will be listed as a [Packer Integration](/packer/integrations), with details on how to install and usage your the plugin. 

> Plugin integration, once deployed, can be updated manually, or automatically upon a new release, by the plugin authors. Changes to the defined documentation structure 

Reviewer: ```suggestion
Plugin integrations, once deployed, can be updated manually, or automatically upon a new release, by the plugin authors. Changes to the defined documentation structure 
```

or parent repository should be communicated to the Packer team to ensure a working integration pipeline. 

The following steps are required for registering your plugin as an integration

> 1. Open a request for integration issue with the Packer team - [LINK]().

Author: I will add the issue template now.

Please provide all the requested information to help expedite the integration request. 
1. Update your plugin documentation structure according to the templated defined with the [Packer Plugin Scaffolding template].
New plugin generated from this template may have the necessary structure in place. If so you can jump to step 4.
1. For the integrations library, only one top-level README per integration is supported. Any top-level index.mdx files that exist
within a plugins existing documentation will need to migrate to a top-level README. 
1. Update your top-level integration README to include a description, plugin installation steps, available components section, and, any, additional sections
needed to inform users on how to work with your integration. Refer to [Packer scaffolding plugin](https://github.com/hashicorp/packer-plugin-scaffolding/blob/main/docs/README.md) for an example. 
1. Update the top-level README for each of the components within your integration to follow the structure defined in the scaffolding template.
1.

> ### Testing Plugin Documentation

Author: internally we have a task for this but we need to circle back to see what we can do.
